examples/NAT.cs

```diff
@@ -36,223 +36,348 @@
     {
-      return Port_Drop;
-    }
-
-    // Unencapsulate the packets, so we can read and change their fields more easily.
-    IpPacket p_ip = ((IpPacket)(packet.PayloadPacket));
-    TcpPacket p_tcp = ((PacketDotNet.TcpPacket)(p_ip.PayloadPacket));
+      if (packet.Encapsulates(typeof(IPv4Packet), typeof(TcpPacket)))
+      {
+        // Unencapsulate the packets, so we can read and change their fields more easily.
+        EthernetPacket p_eth = (EthernetPacket)packet;
+        IpPacket p_ip = ((IpPacket)(packet.PayloadPacket));
+        TcpPacket p_tcp = ((PacketDotNet.TcpPacket)(p_ip.PayloadPacket));
 
-    // Prepare the structure with which we'll query our port mapping.
-    NAT_Entry from = new NAT_Entry(); //FIXME for increased efficiency could avoid this allocation, and use a thread-local but static one done at the start?
-    from.ip_address = p_ip.SourceAddress;
-    from.tcp_port = p_tcp.SourcePort;
+#if DEBUG
+        Console.WriteLine("RX {0}:{1} -> {2}:{3} on {4} [{5}]",
+          p_ip.SourceAddress, p_tcp.SourcePort, p_ip.DestinationAddress, p_tcp.DestinationPort, in_port,
+          p_tcp.Syn?"S":"");
+#endif
 
-    int out_port = Port_Drop;
-    if (in_port == Port_Outside)
-    {
-      from.assigned_tcp_port = p_tcp.DestinationPort;
-      from.network_port = null;
-      out_port = outside_to_inside (p_ip, p_tcp, from);
-    } else {
-      from.assigned_tcp_port = null;
-      from.network_port = in_port;
-      out_port = inside_to_outside (p_ip, p_tcp, from);
-    }
+        int out_port = Port_Drop;
+        if (in_port == Port_Outside)
+          out_port = outside_to_inside (p_eth, p_ip, p_tcp);
+        else
+          out_port = inside_to_outside (p_eth, p_ip, p_tcp, in_port);
 
 #if DEBUG
-    print_mapping (port_mapping, " (------ Outside ------) ", " (------ Inside ------) ");
-    Console.WriteLine();
-    print_mapping (port_reverse_mapping, " (------ Inside ------) ", " (------ Outside ------) ");
+        Console.WriteLine(" (------ Outside ------)  \t<->\t (------ Inside ------) ");
+        foreach (var entry in NAT_MapToInside)
+        {
+          Console.WriteLine("{0} \t<->\t {1}", entry.Key, entry.Value);
+        }
+        Console.WriteLine();
+        Console.WriteLine(" (------ Inside ------)  \t<->\t (------ Outside ------) ");
+        foreach (var entry in NAT_MapToOutside)
+        {
+          Console.WriteLine("{0} \t<->\t {1}", entry.Key, entry.Value);
+        }
 #endif
 
-    return out_port;
+        return out_port;
+      }
+    }
+
+    // Drop packets that aren't handled
+    return Port_Drop;
   }
 
-  private int outside_to_inside (IpPacket p_ip, TcpPacket p_tcp, NAT_Entry from)
+  /// <summary>
+  /// Rewrite packets coming from the Outside and forward on the relevant Inside network port.
+  /// </summary>
+  private int outside_to_inside (EthernetPacket p_eth, IpPacket p_ip, TcpPacket p_tcp)
   {
+    // p_ip.DestinationAddress should be my_address
```

Comment:
Makes sense, thanks.

On 2016-07-07 12:50, TMVector wrote:
> In examples/NAT.cs [1]:
> 
>> {
>> + // p_ip.DestinationAddress should be my_address
> 
> No. I configure all the hosts to use the NAT as a gateway, but packets
> are still addressed to their actual destination.
> 
> --
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub [2], or mute the
> thread [3].
> 
> 	*
> 
> Links:
> ------
> [1] https://github.com/niksu/pax/pull/3#discussion_r69893012
> [2]
> https://github.com/niksu/pax/pull/3/files/d721adebf668133ae20c55f466ccfaa2344f51a2#r69893012
> [3]
> https://github.com/notifications/unsubscribe/ABDTTkGToVWYoiZtC6qCiVW3fK8I_Autks5qTOgfgaJpZM4JFDoJ

-- 
http://www.cl.cam.ac.uk/~ns441/